Resolves: Add native GitHub continuous code security and quality analysis

aleks-ivanov opened the pull request:
- add `codeql-analysis.yml` which automatically enables CodeQL code security and quality scanner. It executes on every push commit, PR, manually and every day at 8:00AM UTC. A scan check can be viewed on commits and PRs. All alerts and fix suggestion can be viewed at **Security** tab -> **Code scanning alerts** -> **CodeQL**

### File changes:

- Switched the places of the directives in `src/Microsoft.AspNetCore.Mvc.Versioning/Versioning/ActionSelectionContext.cs`, because of this error:

```
src\Microsoft.AspNetCore.Mvc.Versioning\Versioning\ActionSelectionContext.cs(5,5): error SA1210: Using directives should be ordered alphabetically by the namespaces
```

Resolves #766

commonsensesoftware commented on .github/workflows/codeql-analysis.yml:
```diff
@@ -0,0 +1,55 @@
+          $pathToSolution = "ApiVersioning.sln"
+          $buildConfiguration = "Release"
+          $useSharedCompilation = "false"
+          $testProjects = @( Get-ChildItem -Path test/ -Recurse | Where-Object { $_.Extension -match 'shproj' -or $_.Extension -match 'csproj' } | ForEach-Object { $_.FullName } )
```
This command can be simplified to just:

```powershell
Get-ChildItem -Path test -Include *.csproj -Recurse | Select-Object FullName
```

A couple of supporting points:

- This _assumes_ (or rather _declares_) it needs to run on Windows, but it doesn't _have_ to be restricted to that (in the future)
- Avoid literal `/` or `\` to ensure cross-platform builds; PowerShell will do the _right_ thing
- There is no need to filter _Shared Projects_ (`*.shproj`) separately because they are automatically included in all C# projects (`*.csproj`). _Shared Projects_ do not compile or have their own output.
- There is no need to wrap the results in an array (`@()`) because the results are already an array. There will always be 2+ items.

commonsensesoftware commented on .github/workflows/codeql-analysis.yml:
```diff
@@ -0,0 +1,55 @@
+          $useSharedCompilation = "false"
+          $testProjects = @( Get-ChildItem -Path test/ -Recurse | Where-Object { $_.Extension -match 'shproj' -or $_.Extension -match 'csproj' } | ForEach-Object { $_.FullName } )
+
+          dotnet nuget locals all --clear
```
What is the value of clearing the NuGet cache? This doesn't appear to do anything useful, except slow down the build. This _might_ be useful if local, self-referencing packages were used from build output, but this project doesn't do that. Packages are immutable with idempotent behavior.

Slowing down the build isn't the _worse_ thing in the world; however, if someone runs this locally, this is pretty significant side effect that a user/dev may not realize will happen. I'm always leery of anything that makes changes to someone's local system.

commonsensesoftware commented on .github/workflows/codeql-analysis.yml:
```diff
@@ -0,0 +1,55 @@
+
+          # remove one or more test projects, 
+          # so that CodeQL only analyzes the source code
+          dotnet sln $pathToSolution remove $testProjects
```
This is a bad idea. This makes modifications to the Solution (`*.sln`) file. Yes, it can be undone with source control or thrown away in a temporary build, but that is not intrinsically obvious to someone testing this locally. There are at least two alternatives.

### Alterative 1 : Loop Over Target Projects
I just realized that the query above is filtering out test projects. Inverting the query would allow looping over the target projects without any modification.  For example:

```powershell
Get-ChildItem -Path src -Include *.csproj -Recurse | ForEach-Object { <# use $_.FullName #> }
```

### Alternative 2 : Use Solution Filters
The simpler and _saner_ approach would be to use [Solution Filters](https://docs.microsoft.com/en-us/visualstudio/msbuild/solution-filters) (`*.slnf`). I presume this would be called something like `codeql-analysis.slnf` and only include the projects to be analyzed. All `dotnet` commands would then accept the `*.slnf` file instead of the `*.sln` file.

> NOTE: `*.slnf` files have been natively supported by MSBuild since 16.7 (see [Issue 4097](https://github.com/dotnet/msbuild/issues/4097))

commonsensesoftware commented on .github/workflows/codeql-analysis.yml:
```diff
@@ -0,0 +1,55 @@
+          -property:UseSharedCompilation=$useSharedCompilation `
+          --configuration $buildConfiguration `
+          --no-incremental `
+          --no-restore
```
Why break the restore process into two different steps? `dotnet build` intrinsically triggers restore unless you specify `--no-restore` as you've done. Removing `--no-restore` and the `dotnet restore` will merge the operations into one. In addition, the `dotnet restore` command does **not** specify the `--configuration` parameter. As unlikely as it is, it is possible that restore behavior could vary between configurations. If it did, these two commands could produce a failure. That is negated by either using `dotnet restore --configuration $buildConfiguration` or merging these two commands into one.

commonsensesoftware commented on .github/workflows/codeql-analysis.yml:
```diff
@@ -0,0 +1,55 @@
+      - name: Perform CodeQL Analysis
+        uses: github/codeql-action/analyze@v1
+
+# Built with ❤ by [Pipeline Foundation](https://pipeline.foundation)
```
nit: Curious if this is injected or code-generated? I'm not opposed to _advertising_ things that drive you to the _Pit of Success_. 😉

aleks-ivanov replied:
This is simply showing our contribution to the project. We'd really appreciate it if you support our non-profit initiative - Pipeline Foundation 🙂

commonsensesoftware reviewed (commented):
Overall, looks good. A few basic thoughts and feedback points.

The change to **ActionSelectionContext.cs** is fine. Not sure why code analysis didn't catch this before. I've notice that the results are occasionally miss some items.

aleks-ivanov commented:
@commonsensesoftware on top of the other suggested improvements, after testing the suggested changes in the [dev container thread](https://github.com/dotnet/aspnet-api-versioning/issues/763) (the capitalization of `Directory.Build.props` and `Directory.Build.targets`.) and ensuring the source projects are being built without errors on Ubuntu 20.04, we implemented those fixes here and switched the pipeline runner from Windows to Ubuntu for increased speed of execution.
